[ETHEREUM-CONTRACTS] New web3-operations path for SuperTokenFactory

0xdavinchee opened the pull request:
## Context

This PR originally began to solve the issue of being unable to add arbitrary update logic within the SuperTokenFactory-namely the deployment of the SuperToken and the NFT logic contracts. However, after a review, it was made clear the existing flow was a bit confusing.

## Problem
- Unintuitive ergonomics and hidden deployments when updating SuperTokenFactory logic contract
  - Side effect of: deploying new super token logic contract and setting it as the canonical super token logic in super token factory
  - Note: There is tight coupling between super token factory and super token currently and in the future it will be the same. It is therefore important to update both the super token factory logic contract and the super token logic for all super tokens otherwise the newly deployed super token's may have different logic than existing ones.

## Solution

Previous SuperTokenFactory Upgrade Flow:
- Deploy new SuperTokenFactoryHelper: `helper`
- Deploy new SuperTokenFactory(host, `helper`): `superTokenFactoryNewLogic`
- call governance.updateContracts(host, hostNewLogic, newAgreementClassLogics, `superTokenFactoryNewLogic`)
  - host.updateSuperTokenFactory(`superTokenFactoryNewLogic`)
  - superTokenFactoryProxy.updateCode(`superTokenFactoryNewLogic`)
  - _updateCodeAddress: updates super token factory logic contract address to `superTokenFactoryNewLogic`
  - _updateSuperTokenLogic: deploys new super token logic contract, castrates it and sets it as canonical new super token logic address in super token factory

New SuperTokenFactory Upgrade Flow:
- Deploy new SuperToken logic: `newSuperTokenLogic`
- Deploy new SuperTokenFactory(host, `newSuperTokenLogic`): `superTokenFactoryNewLogic`
  - This sets the new immutable `_superTokenLogic` field as `newSuperTokenLogic` and also castrates the logic contract 
    - > IMPORTANT CAVEAT: On the next upgrade to the contract in this PR, the Super Token Logic will not castrate the logic contract in the constructor because the previous logic contract does this in the `_updateSuperTokenLogic` call
- call governance.updateContracts(host, hostNewLogic, newAgreementClassLogics, `superTokenFactoryNewLogic`)
  - host.updateSuperTokenFactory(`superTokenFactoryNewLogic`)
  - superTokenFactoryProxy.updateCode(`superTokenFactoryNewLogic`)
  - _updateCodeAddress: updates super token factory logic contract address to `superTokenFactoryNewLogic`

0xdavinchee committed:
add new updateLogicContracts

0xdavinchee committed:
add sanity check test

- add mock contract
- add test to ensure new code is being run

0xdavinchee committed:
forge install: forge-std

v1.3.0

0xdavinchee committed:
SuperTokenFactory deploy test

hellwolf commented:
Let's split some mini factories out of the main token factory

0xdavinchee committed:
fix framework test

0xdavinchee committed:
forge install: forge-std

v1.3.0

0xdavinchee commented on packages/ethereum-contracts/ops-scripts/deploy-framework.js:
```diff
@@ -577,6 +579,30 @@ module.exports = eval(`(${S.toString()})({skipArgv: true})`)(async function (
             }
         },
         async () => {
+            const deploySuperTokenDeployerLibraryAndLinkToFactoryHelper =
```
abstract to HOF

hellwolf commented on packages/ethereum-contracts/dev-scripts/deploy-test-framework.js:
```diff
@@ -6,6 +6,7 @@ const SuperfluidCFAv1DeployerLibraryArtifact = require("@superfluid-finance/ethe
 const SuperfluidIDAv1DeployerLibraryArtifact = require("@superfluid-finance/ethereum-contracts/artifacts/contracts/utils/deployers/SuperfluidIDAv1DeployerLibrary.sol/SuperfluidIDAv1DeployerLibrary.json");
 const SuperfluidPeripheryDeployerLibraryArtifact = require("@superfluid-finance/ethereum-contracts/artifacts/contracts/utils/deployers/SuperfluidPeripheryDeployerLibrary.sol/SuperfluidPeripheryDeployerLibrary.json");
 const SuperfluidSuperTokenFactoryHelperDeployerLibraryArtifact = require("@superfluid-finance/ethereum-contracts/artifacts/contracts/utils/deployers/SuperfluidSuperTokenFactoryHelperDeployerLibrary.sol/SuperfluidSuperTokenFactoryHelperDeployerLibrary.json");
```
SuperfluidSuperToken is a bit redundant naming

0xdavinchee committed:
fix redundant naming

0xdavinchee committed:
Ops flow cleanup

- delete updateLogicContracts() function
- delete SuperTokenFactoryHelper
- delete _updateSuperTokenLogic function (future updates will no longer follow this path)
- use SuperTokenDeployerLibrary.deploySuperTokenLogic to deploy non upgradeable super token
- fix breaking tests
- create some abstracted functions for deploy external library and linking

0xdavinchee commented on packages/ethereum-contracts/contracts/superfluid/SuperTokenFactory.sol:
```diff
@@ -32,9 +29,14 @@ abstract contract SuperTokenFactoryBase is
+    // @dev This is the old SuperToken logic contract that is no longer used
+    // It is kept here for backwards compatibility due to the fact that we cannot
+    // change the storage layout of the contract
+    ISuperToken internal _superTokenLogicDeprecated;
```
NOTE: we deprecate the previous `_superTokenLogic` field in storage in favor of the immutable field which will be updated on each subsequent deployment of a new `SuperTokenFactory` logic contract.

0xdavinchee commented on packages/ethereum-contracts/contracts/superfluid/SuperTokenFactory.sol:
```diff
@@ -32,9 +29,14 @@ abstract contract SuperTokenFactoryBase is
         variables are added APPEND-ONLY. Re-ordering variables can
         permanently BREAK the deployed proxy contract. */
 
+    ISuperToken immutable public _superTokenLogic;
```
NOTE: the "new" immutable `_superTokenLogic` variable.

hellwolf replied:
```
packages/ethereum-contracts/contracts/superfluid/Superfluid.sol:    bool immutable public NON_UPGRADABLE_DEPLOYMENT;
```

Be consistent re using snake cases for immutables.

0xdavinchee replied:
addressed: 674e8fbfd2b6cb5a5fb28df6c8b33c97e7c034f4

0xdavinchee commented on packages/ethereum-contracts/contracts/superfluid/SuperTokenFactory.sol:
```diff
@@ -49,9 +51,24 @@ abstract contract SuperTokenFactoryBase is
+        // it will revert the first time, however we MUST uncomment it after subsequent 
+        // updates to the logic contract so that the super token logic contract is initialized
+        // here 
+        // UUPSProxiable(address(_superTokenLogic)).castrate();
```
**CAUTION: **We leave the castrate function commented out for this upcoming upgrade-please read comment above in full.

In the ERC20x PR we **MUST** uncomment this line so that the super token logic contract is initialized in the constructor.

hellwolf replied:
"castrate step".

0xdavinchee commented on packages/ethereum-contracts/contracts/superfluid/SuperTokenFactory.sol:
```diff
@@ -83,14 +102,6 @@ abstract contract SuperTokenFactoryBase is
-        _updateSuperTokenLogic();
-    }
-
-    function _updateSuperTokenLogic() private {
```
NOTE: We delete this as `_superTokenLogic` is being set in the constructor as an immutable field.

0xdavinchee commented on packages/ethereum-contracts/contracts/superfluid/SuperTokenFactory.sol:
```diff
@@ -175,7 +186,7 @@ abstract contract SuperTokenFactoryBase is
         }
 
         if (upgradability == Upgradability.NON_UPGRADABLE) {
-            superToken = ISuperToken(this.createSuperTokenLogic(_host));
```
NOTE: We use our linked `SuperTokenDeployerLibrary` to deploy the super token logic here now instead.

hellwolf replied:
?

0xdavinchee replied:
this no longer applies as SuperToken logic is deployed outside of the SuperTokenFactory contract

0xdavinchee commented on packages/ethereum-contracts/contracts/superfluid/SuperTokenFactory.sol:
```diff
@@ -298,15 +309,6 @@ abstract contract SuperTokenFactoryBase is
     }
 }
 
-// splitting this off because the contract is getting bigger
```
NOTE: the SuperTokenFactoryHelper is removed

0xdavinchee commented on packages/ethereum-contracts/contracts/superfluid/SuperTokenFactory.sol:
```diff
@@ -315,22 +317,25 @@ contract SuperTokenFactory is SuperTokenFactoryBase
     constructor(
         ISuperfluid host,
-        SuperTokenFactoryHelper helper
+        ISuperToken superTokenLogic
```
NOTE: We no longer pass in the helper in the `SuperTokenFactory` constructor, we deploy a super token logic contract and then pass it into the constructor of the SuperTokenFactory, which will castrate it.

hellwolf replied:
ops-script should be fixed.

0xdavinchee commented on packages/ethereum-contracts/contracts/superfluid/SuperTokenFactory.sol:
```diff
@@ -315,22 +317,25 @@ contract SuperTokenFactory is SuperTokenFactoryBase
     }
 
-    function createSuperTokenLogic(ISuperfluid host)
+    /// DEPRECATED
```
NOTE: this is deprecated and will be deleted in the next PR

0xdavinchee commented on packages/ethereum-contracts/ops-scripts/deploy-framework.js:
```diff
@@ -386,26 +384,35 @@ module.exports = eval(`(${S.toString()})({skipArgv: true})`)(async function (
         )(superfluid.address, cfa.address);
     }
 
+    const deployExternalLibraryAndLink = async (
```
NOTE: this is an abstracted function used for:
- deploying an external library
- linking a contract artifact to the deployed external library (in two ways depending on if hardhat or truffle)
- returning the deployed external library

hellwolf replied:
convert it to comments instead.

hellwolf replied:
Will need to manually test "hot-fuzz" for any external library linkage changes: https://github.com/superfluid-finance/protocol-monorepo/blob/master/packages/hot-fuzz/index.js

0xdavinchee replied:
addressed: b81c954ac1b805caf1435db0f98cb2dba8d8ad3e

0xdavinchee commented on packages/ethereum-contracts/test/contracts/superfluid/SuperTokenFactory.test.ts:
```diff
@@ -147,14 +168,16 @@ describe("SuperTokenFactory Contract", function () {
                 );
                 await updateSuperTokenFactory();
-                assert.equal((await superToken1.waterMark()).toString(), "0");
+                // @note I am not sure what the original intention of waterMark is
```
NOTE: I dug into the js-sdk code and I'm not sure why this call to get waterMark fails only in the case when upgradability is set to 0.

0xdavinchee committed:
cleanup

0xdavinchee reviewed (commented):
Some parts to pay extra attention to

0xdavinchee committed:
Merge branch 'dev' into supertokenfactory-new-update

0xdavinchee committed:
Merge branch 'supertokenfactory-new-update' of github.com:superfluid-finance/protocol-monorepo into supertokenfactory-new-update

0xdavinchee committed:
fork baseline

- add fork baseline tests file
- rename fork super token factory upgrade test file

0xdavinchee committed:
fork baseline wip

- refactor ForkBaseline: it is responsible for loading the forked framework
- add more tests in ForkBaseline
- do not run fork tests in the CI

0xdavinchee committed:
Update CHANGELOG.md

0xdavinchee committed:
Merge branch 'dev' into supertokenfactory-new-update

0xdavinchee committed:
deprecate non upgradeable deploy

0xdavinchee committed:
merge dev

0xdavinchee committed:
fix broken build

0xdavinchee commented on packages/ethereum-contracts/contracts/superfluid/SuperTokenFactory.sol:
```diff
@@ -175,7 +185,7 @@ abstract contract SuperTokenFactoryBase is
 
         if (upgradability == Upgradability.NON_UPGRADABLE) {
-            superToken = ISuperToken(this.createSuperTokenLogic(_host));
+            revert SUPER_TOKEN_FACTORY_NON_UPGRADEABLE_IS_DEPRECATED();
```
NOTE: non upgradeable super token deployment via SuperTokenFactory is deprecated now.

hellwolf commented on .github/workflows/ci.feature.yml:
```diff
@@ -79,6 +79,8 @@ jobs:
         run: |
           yarn workspace @superfluid-finance/ethereum-contracts test
+        env:
+          POLYGON_MAINNET_PROVIDER_URL: ${{ secrets.POLYGON_MAINNET_PROVIDER_URL }}
```
`POLYGON_MAINNET_ARCHIVE_PROVIDER_URL`

0xdavinchee replied:
addressed a35994de537596d86ed83659c8c54c3dcbd29089

hellwolf commented on packages/ethereum-contracts/contracts/utils/SuperfluidFrameworkDeployer.sol:
```diff
@@ -15,8 +15,8 @@ import {
     SuperfluidIDAv1DeployerLibrary
 } from "./deployers/SuperfluidIDAv1DeployerLibrary.sol";
```
merge "deployers/*.sol" here.

0xdavinchee replied:
addressed: 9e44565721a174d7b0e99efd7e4c74fe65dae46e

hellwolf commented on packages/ethereum-contracts/test/foundry/deployments/ForkBaseline.t.sol:
```diff
@@ -0,0 +1,632 @@
+// SPDX-License-Identifier: AGPLv3
```
Test vector management for future reference:

-  This is more like a smoke test.
-  Think of how to merge into ethereum-contracts/test/foundry.

0xdavinchee replied:
addressed: 36e0b2e361467d1b06d8d659ecd291f977c46d38

hellwolf commented on packages/ethereum-contracts/test/foundry/deployments/ForkSuperTokenFactoryUpgrade.t.sol:
```diff
@@ -0,0 +1,170 @@
+    using SuperTokenV1Library for ISuperToken;
+    string public PROVIDER_URL;
+
+    IResolver public constant resolver =
```
It is not very clear from the context that this is for Polygon only.

0xdavinchee replied:
addressed: f22d6aee070fdeba4d2c4d6e0144bb4272dfeefd

hellwolf reviewed (changes requested):
please address the comments.

0xdavinchee committed:
Address CI env naming comment

https://github.com/superfluid-finance/protocol-monorepo/pull/1276#discussion_r1104197334

0xdavinchee committed:
address merge "deployers/*.sol" comment

https://github.com/superfluid-finance/protocol-monorepo/pull/1276#discussion_r1104220890

0xdavinchee committed:
use consistent immutable var naming

https://github.com/superfluid-finance/protocol-monorepo/pull/1276#discussion_r1104205341

0xdavinchee committed:
address documentation comment

https://github.com/superfluid-finance/protocol-monorepo/pull/1276#discussion_r1104225932

0xdavinchee committed:
address scripts artifacts naming

change from scripts -> ops-scripts as the test/ops-scripts are testing ops-scripts

0xdavinchee committed:
address baseline comment

https://github.com/superfluid-finance/protocol-monorepo/pull/1276#discussion_r1104241091

0xdavinchee committed:
address missing context naming comment

https://github.com/superfluid-finance/protocol-monorepo/pull/1276#discussion_r1104244941

0xdavinchee committed:
Merge branch 'dev' into supertokenfactory-new-update

0xdavinchee committed:
cleanup

- change hot-fuzz to match new file format
- remove unused UUPSProxy import

d10r commented on .github/workflows/ci.feature.yml:
```diff
@@ -79,6 +79,9 @@ jobs:
         run: |
           yarn workspace @superfluid-finance/ethereum-contracts test
+        env:
+          # NOTE: This is currently unset and fork tests are not being run
```
I have set the ENV var POLYGON_MAINNET_ARCHIVE_PROVIDER_URL

0xdavinchee committed:
duplicate keys makes no sense

ty Miao

d10r commented on packages/ethereum-contracts/ops-scripts/deploy-framework.js:
```diff
@@ -578,14 +591,19 @@ module.exports = eval(`(${S.toString()})({skipArgv: true})`)(async function (
-                SuperTokenFactoryHelperLogic.new,
-                "SuperTokenFactoryHelperLogic.new"
-            )();
+            const superTokenLogic = useMocks
+                ? await web3tx(SuperTokenLogic.new, "SuperTokenLogic.new")(
+                      superfluid.address,
+                      0
+                  )
+                : await web3tx(
+                      SuperTokenLogic.new,
+                      "SuperTokenLogic.new"
+                  )(superfluid.address);
```
I guess the way this is structured, the script would always deploy a new instance of SuperTokenLogic, even if only the factory code changed.
Wouldn't break anything, but potentially waste gas.
I suggest we leave it like this for now and unbundle when modifying further for NFT deployment.

hellwolf replied:
let's do the fine tuning in ERC20x branch.

0xdavinchee commented on packages/ethereum-contracts/contracts/superfluid/SuperTokenFactory.sol:
```diff
@@ -49,9 +59,24 @@ abstract contract SuperTokenFactoryBase is
+        // and passed in as a parameter to SuperTokenFactory constructor
+        _SUPER_TOKEN_LOGIC = superTokenLogic;
+        
+        // @note this function call is commented out on the first upgrade
```
Signpost for ERC20x
